getdnsapi/stubby
# Add Dockerfile

Enables development and testing with just Docker installed. Also, it can be used to deploy in production, although it's not only optimized for that usage (eg. image size). Useful for building and running the latest HEAD (or whatever branch, tag or commit id you specify) straight from source by just issuing a one-liner (see below) without any requirement other than Docker. Dockerfile is commented and contains an example run command.

Image builds code on latest Debian LTS (Bullseye), with these library versions:
- libidn2 2.3.0
- libgetdns 1.6.0
- libunbound 1.13.1
- openssl 1.1.1k

The built binary and the example configuration file are copied to the final image. The configuration file gets added the `0.0.0.0` listen address to allow the program to get incoming connections from outside the container. I skipped IPv6, since I suspect I was being hit by #281 or similar and once by #18, IIRC and therefore I could not test it reliably. If IPv6 it's required, please advise.

Notice that `stubby` will be forcefully killed on `docker stop` after timeout due to #188.

To test the build locally, issue the following:
``` sh
docker build --force-rm --pull github.com/pataquets/stubby#dockerfile -t stubby
```
See Dockerfile for usage tips and how to launch a test container. Just add `-l` at the command's end to debug.

Attached below is a partial `docker build` log with just the compilation steps, for reviewing:
```
Step 6/15 : RUN cmake .
 ---> Running in c173bb614500
-- No build type defined; defaulting to 'Debug'
-- The C compiler identification is GNU 10.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Could NOT find Libsystemd (missing: LIBSYSTEMD_LIBRARIES LIBSYSTEMD_INCLUDE_DIR)
-- Found Libyaml: /usr/lib/x86_64-linux-gnu/libyaml.so
-- Looking for os/log.h
-- Looking for os/log.h - not found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for getopt
-- Looking for getopt - found
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found suitable version "1.1.1k", minimum required is "1.0.2")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
-- Found Libidn2: /usr/lib/x86_64-linux-gnu/libidn2.so (found suitable version "2.3.0", minimum required is "2.0.0")
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.1.1k")
-- Found Libunbound: -pthread;/usr/lib/x86_64-linux-gnu/libssl.so;/usr/lib/x86_64-linux-gnu/libcrypto.so;/usr/lib/x86_64-linux-gnu/libunbound.so (found suitable version "1.13.1", minimum required is "1.5.9")
-- Found Getdns: OpenSSL::SSL;OpenSSL::Crypto;Threads::Threads;Libidn2::Libidn2;Libunbound::Libunbound;/usr/lib/x86_64-linux-gnu/libgetdns.so (found suitable version "1.6.0", minimum required is "1.5.0")
-- Configuring done
-- Generating done
-- Build files have been written to: /usr/src/stubby
Removing intermediate container c173bb614500
 ---> b5150f6c4221
Step 7/15 : RUN make
 ---> Running in f091dbff6dcf
Scanning dependencies of target stubby
[ 12%] Building C object CMakeFiles/stubby.dir/src/stubby.c.o
[ 25%] Building C object CMakeFiles/stubby.dir/src/configfile.c.o
[ 37%] Building C object CMakeFiles/stubby.dir/src/log.c.o
[ 50%] Building C object CMakeFiles/stubby.dir/src/server.c.o
[ 62%] Building C object CMakeFiles/stubby.dir/src/util.c.o
[ 75%] Building C object CMakeFiles/stubby.dir/src/yaml/convert_yaml_to_json.c.o
[ 87%] Building C object CMakeFiles/stubby.dir/src/sldns/sbuffer.c.o
[100%] Linking C executable stubby
[100%] Built target stubby
Removing intermediate container f091dbff6dcf
```
